public/js/actions/sources.js

```diff
@@ -95,29 +86,30 @@ function newSources(sources) {
  * @static
  */
 function loadSourceMap(generatedSource) {
-  return ({ dispatch, getState }) => {
-    let sourceMap = getSourceMap(getState(), generatedSource.id);
-    if (sourceMap) {
+  return async function({ dispatch, getState }) {
+    const map = await fetchSourceMap(generatedSource);
+    if (!map) {
+      // If this source doesn't have a sourcemap, do nothing.
       return;
     }
 
-    dispatch({
-      type: constants.LOAD_SOURCE_MAP,
-      source: generatedSource,
-      [PROMISE]: (async function () {
-        const sourceMapURL = getSourceMapURL(getState(), generatedSource);
-        sourceMap = await networkRequest(sourceMapURL);
-
-        const originalSources = await createOriginalSources(
-          generatedSource,
-          sourceMap
-        );
+    const originalSources = map.sources.map(originalUrl => {
+      return {
+        url: originalUrl,
+        id: generatedToOriginalId(generatedSource.id, originalUrl),
+        isPrettyPrinted: false
+      };
+    });
 
-        originalSources.forEach(s => dispatch(newSource(s)));
+    originalSources.forEach(s => dispatch(newSource(s)));
 
-        return { sourceMap };
-      })()
+    dispatch({
+      type: constants.SOURCE_MAP_LOADED,
+      source: generatedSource,
+      map: map
     });
+
+    return map;
```

Comment:
It's not, but usually async actions return a meaningful object. With the promise middleware they always return what the promise returns. So it's just a consistent API that might allow for things we can't think of yet.